File: tests/gizclaw-e2e/flutter/giztest/lib/src/timing.dart
/// The Go/C load runner executes scheduling. This SDK contract runner validates
/// and ignores these fields and records timing_mode: ignored in its report.
library;

final _pattern = RegExp(r'^(0|([0-9]+(\.[0-9]+)?(ns|us|µs|μs|ms|s|m|h))+)$');
final _part = RegExp(r'([0-9]+)(?:\.([0-9]+))?(ns|us|µs|μs|ms|s|m|h)');
final _maximum = BigInt.parse('9223372036854775807');
const _units = {
  'h': 3600000000000,
  'm': 60000000000,
  's': 1000000000,
  'ms': 1000000,
  'us': 1000,
  'µs': 1000,
  'μs': 1000,
  'ns': 1,
};

BigInt _duration(Object? value, String field) {
  if (value is! String || !_pattern.hasMatch(value)) {
    throw FormatException('$field must be a non-negative duration');
  }
  var total = BigInt.zero;
  for (final match in _part.allMatches(value)) {
    final fraction = match.group(2) ?? '';
    final scale = BigInt.from(10).pow(fraction.length);
    total +=
        BigInt.parse('${match.group(1)}$fraction') *
        BigInt.from(_units[match.group(3)]!) ~/
        scale;
  }
  if (total > _maximum) {
    throw FormatException('$field exceeds the duration range');
  }
  return total;
}

void validateTiming(Map<String, Object?> fields) {
  BigInt duration(String name) =>
      fields.containsKey(name) ? _duration(fields[name], name) : BigInt.zero;
  final start = duration('start_jitter');
  final stagger = duration('stagger');
  final step = duration('step_jitter');
  final seed = fields['seed'];
  if (fields.containsKey('seed') &&
      (seed is! int || seed < 0 || seed > 9007199254740991)) {
    throw const FormatException(
      'seed must be an integer in 0..9007199254740991',
    );
  }
  final repeat = fields['repeat'] ?? 1;
  if (repeat is int &&
      repeat > 0 &&
      BigInt.from(repeat - 1) * stagger +
              (start > BigInt.zero ? start - BigInt.one : BigInt.zero) >
          _maximum) {
    throw const FormatException(
      '(repeat - 1) * stagger + start_jitter exceeds the duration range',
    );
  }
  final steps = fields['steps'];
  if (steps is List &&
      steps.any((step) => step is Map && step['barrier'] != null) &&
      (start > BigInt.zero || stagger > BigInt.zero || step > BigInt.zero)) {
    throw const FormatException(
      'barrier cannot be combined with non-zero start_jitter, stagger or step_jitter',
    );
  }
}

File: tests/gizclaw-e2e/js/giztest/timing.ts
// Scheduling is executed by the shared Go/C runner. This SDK contract runner
// validates the same fields before accepting and ignoring them; its report
// marks timing_mode as "ignored" so it cannot be mistaken for a load test.
const durationPattern = /^(0|([0-9]+(\.[0-9]+)?(ns|us|µs|μs|ms|s|m|h))+)$/u;
const maxDuration = 9_223_372_036_854_775_807n;
const units: Record<string, bigint> = {
  h: 3_600_000_000_000n,
  m: 60_000_000_000n,
  ms: 1_000_000n,
  ns: 1n,
  s: 1_000_000_000n,
  us: 1000n,
  µs: 1000n,
  μs: 1000n,
};

function duration(value: unknown, field: string): bigint {
  if (typeof value !== "string" || !durationPattern.test(value)) {
    throw new Error(`${field} must be a non-negative duration`);
  }
  let total = 0n;
  for (const match of value.matchAll(
    /([0-9]+)(?:\.([0-9]+))?(ns|us|µs|μs|ms|s|m|h)/gu,
  )) {
    const fraction = match[2] ?? "";
    const scale = 10n ** BigInt(fraction.length);
    total += (BigInt(match[1] + fraction) * units[match[3]]) / scale;
  }
  if (total > maxDuration) {
    throw new Error(`${field} exceeds the duration range`);
  }
  return total;
}

export type TimingFields = {
  start_jitter?: unknown;
  stagger?: unknown;
  step_jitter?: unknown;
  seed?: unknown;
};

export function validateTiming(
  fields: TimingFields,
  repeat: number,
  hasBarrier: boolean,
): void {
  const start =
    fields.start_jitter === undefined
      ? 0n
      : duration(fields.start_jitter, "start_jitter");
  const stagger =
    fields.stagger === undefined ? 0n : duration(fields.stagger, "stagger");
  const step =
    fields.step_jitter === undefined
      ? 0n
      : duration(fields.step_jitter, "step_jitter");
  if (
    fields.seed !== undefined &&
    (typeof fields.seed !== "number" ||
      !Number.isSafeInteger(fields.seed) ||
      fields.seed < 0)
  ) {
    throw new Error("seed must be an integer in 0..9007199254740991");
  }
  if (
    Number.isInteger(repeat) &&
    BigInt(Math.max(0, repeat - 1)) * stagger + (start > 0n ? start - 1n : 0n) >
      maxDuration
  ) {
    throw new Error(
      "(repeat - 1) * stagger + start_jitter exceeds the duration range",
    );
  }
  if (hasBarrier && (start > 0n || stagger > 0n || step > 0n)) {
    throw new Error(
      "barrier cannot be combined with non-zero start_jitter, stagger or step_jitter",
    );
  }
}
